urldeb33fd2fa_friedemann_distripedia_tgz-pJ8-org_json_TestJ8/src/grinder/controller/Controller.java

```diff
@@ -0,0 +1,346 @@
+
+	private ArrayList<String> articleList;
+
+	private Properties props;
+
+	/**
+	 * Instantiates a new controller and starts the server thread
+	 */
+	public Controller(final ControllerGUI cgui) {
+
+		articleList = null;
+		this.cgui = cgui;
+
+		loadProperties("controller.properties");
+		cgui.logMessage("controller initialized, starting server thread...");
+
+		cServer = new CtrlServer(this, GrinderSettings.CONTROLLER_DEFAULT_PORT);
+		new Thread(cServer, "CtrlServer").start();
+	}
+
+	/**
+	 * Load properties.
+	 *
+	 * @param propFileName the prop file name
+	 * @return the properties
+	 */
+	private Properties loadProperties(final String propFileName) {
+
+		final Properties p = new Properties();
+
+		try
+		{
+			final FileInputStream stream = new FileInputStream(propFileName);
+			p.load(stream);
+			stream.close();
+
+			props = p;
+
+		} catch (final FileNotFoundException e)
+		{
+			cgui.logMessage("preset properties file not found - skipping...");
+		} catch (final IOException e)
+		{
+			cgui.logMessage("property reading in failure - skipping...");
+		}
+
+		return p;
+	}
+
+	/**
+	 * @return the props
+	 */
+	public Properties getProps() {
+		return props;
+	}
+
+	/**
+	 * Generates test ticket with user values distributed for connected clients.
+	 *
+	 * @return the test ticket
+	 * @throws URISyntaxException
+	 */
+	private TestTicket generateTestTicket(final int clientsConnected) throws URISyntaxException {
+
+		final TestTicket ticket = cgui.getTestTicketCanvas();
+
+		if (clientsConnected >= 1) // distribute users
+		{
+			ticket.setUserCountHTTP(ticket.getUserCountHTTP() / clientsConnected);
+			ticket.setUserCountWebSocket(ticket.getUserCountWebSocket() / clientsConnected);
+			ticket.setUserCountStratus(ticket.getUserCountStratus() / clientsConnected);
+		}
+
+		ticket.setArticleList(articleList);
+
+		return ticket;
+	}
+
+	/**
+	 * Deploy test tickets to connected clients.
+	 *
+	 * @return true, if successful, false if there are no clients!
+	 */
+	public boolean deployTestTickets() {
+
+		final int clientsConnected = cServer.getServerThreads().size();
+
+		TestTicket t = null;
+		try
+		{
+			t = generateTestTicket(clientsConnected);
+		} catch (final URISyntaxException e)
+		{
+			cgui.logMessage("Webserver address malformed: " + e);
+			return false;
+		}
+
+		if (clientsConnected > 0)
+		{
+			for (final CtrlServerThread cst : cServer.getServerThreads())
+			{
+				cst.writeToOutputStream(ClientCommand.RECEIVE_TESTTICKET);
+				cst.writeToOutputStream(t);
+			}
+			return true;
+		} else
+			return false;
+
+	}
+
+	/**
+	 * Parses the article list into a string array.
+	 *
+	 * @param fileName
+	 * @return int - number of parsed lines/articles, negative if article list
+	 *         was wrongly formatted
+	 * @throws FileNotFoundException
+	 */
+	public int parseArticleList(final String filePath) throws FileNotFoundException {
+
+		final BufferedInputStream fileStream = new BufferedInputStream(new FileInputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fileStream*. Execution paths that do not contain closure statements were detected. Ensure *fileStream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).